k2pow-service/src/job_manager.rs

```diff
@@ -0,0 +1,213 @@
+    TooManyJobs,
+}
+
+/// JobStatus is used to expose job state to external callers
+#[derive(Clone, Debug, PartialEq)]
+pub enum JobStatus {
+    Created,
+    InProgress,
+    Done(Result<u64, String>),
+}
+
+#[derive(Debug)]
+enum JobState {
+    InProgress(std::thread::JoinHandle<Result<u64, post::pow::Error>>),
+    Done(Result<u64, String>),
+}
+
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+pub struct Job {
+    pub nonce_group: u8,
+    pub challenge: [u8; 8],
+    pub difficulty: [u8; 32],
+    pub miner: [u8; 32],
+}
+
+#[cfg_attr(test, mockall::automock)]
+pub trait GetOrCreate {
+    fn get_or_create(&self, job: Job) -> Result<JobStatus, JobError>;
+}
+pub struct JobManager {
+    jobs: Mutex<HashMap<Job, JobState>>,
+    cores: u8,
+    randomx_mode: crate::RandomXMode,
+    randomx_large_pages: bool,
+}
+
+impl JobManager {
+    pub fn new(cores: u8, randomx_mode: crate::RandomXMode, randomx_large_pages: bool) -> Self {
+        JobManager {
+            jobs: Mutex::new(HashMap::new()),
+            cores,
+            randomx_mode,
+            randomx_large_pages,
+        }
+    }
+    fn check_finished(&self, job: &Job) {
+        // there's probably a better way to write this with interior mutability on the
+        // value that the HashMap stores (like with a RefCell). For now, because the
+        // joining the handle consumes the value, borrowing doesn't work here and we need
+        // to remove the value, use it, then put it back into the map.
+        let mut hs = self.jobs.lock().unwrap();
+        let entry = hs.remove_entry(job);
+        match entry {
+            Some((key, JobState::InProgress(handle))) => {
+                if handle.is_finished() {
+                    let val = match handle.join() {
+                        Ok(result) => JobState::Done(match result {
+                            Ok(v) => Ok(v),
+                            Err(e) => Err(e.to_string()),
+                        }),
+                        Err(e) => std::panic::resume_unwind(e),
+                    };
+                    hs.insert(key, val);
+                    return;
+                }
+                hs.insert(key, JobState::InProgress(handle));
+            }
+            Some((key, JobState::Done(res))) => {
+                hs.insert(key, JobState::Done(res));
+            }
+            None => (),
+        }
+    }
+}
+
+impl GetOrCreate for JobManager {
+    fn get_or_create(&self, job: Job) -> Result<JobStatus, JobError> {
+        self.check_finished(&job);
+        let mut hs = self.jobs.lock().unwrap();
+        match hs.get(&job) {
+            Some(JobState::InProgress(_)) => Ok(JobStatus::InProgress),
+            Some(JobState::Done(result)) => Ok(JobStatus::Done(result.clone())),
+            None => {
+                let other = hs.values().find(|v| matches!(v, JobState::InProgress(_)));
+                if other.is_some() {
+                    return Err(JobError::TooManyJobs);
+                }
```

Comment:
Perhaps it could hold also the `Job` that is currently processed to avoid linear search here? I mean something like:

```rust
struct Jobs {
  in_progress: Option<Job>,
  states: HashMap<Job, JobState>,
}

pub struct JobManager {
    jobs: Mutex<Jobs>,
```

Then this becomes:
```suggestion
                if hs.in_progress.is_some() {
                    return Err(JobError::TooManyJobs);
                }
```